git/index/base.py

```diff
@@ -569,16 +569,23 @@ def _preprocess_add_items(self, items):
         """ Split the items into two lists of path strings and BaseEntries. """
         paths = []
         entries = []
-
-        for item in items:
-            if isinstance(item, string_types):
-                paths.append(self._to_relative_path(item))
-            elif isinstance(item, (Blob, Submodule)):
-                entries.append(BaseIndexEntry.from_blob(item))
-            elif isinstance(item, BaseIndexEntry):
-                entries.append(item)
-            else:
-                raise TypeError("Invalid Type: %r" % item)
+        
+        if isinstance(items, string_types):
```

Comment:
I think the repetition can and should be removed. My suggestion is to check if the input type is something you can iterate, and if not, pack it into a list. The rest of the code would then remain unchanged.
Does that make sense?